### ISSUE TITLE
Dockerfile improvements

### ISSUE THREAD

**Issue by frankli0324** [CTFd/CTFd#1781]
I recently see that there are some improvements to the Dockerfile, but there's a little more to do

#### 1. requirements.txt and caching

not only the main requirements.txt can be cached, but also the plugin requirements. this can be done with a "multi-stage" build.

```Dockerfile
# stage 1
FROM python:3.7-slim-buster
COPY . /somewhere
RUN <remove everything let alone requirements.txt>

# stage 2
FROM python:3.7-slim-buster
COPY --from=0 /somewhere .
# now there are only requirements.txt in the folder, and will be cached on next build.
```

#### 2. python:3.7-slim-buster and python3-dev

the image `python:3.7-slim-buster` doesn't contain a python3 package

```cli
root@7d5d3c29ec30:/# apt list --installed | grep python

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

root@7d5d3c29ec30:/# 
```

it copies a custom built python binary into the image instead. you can see that in the dockerhub build history

![image](https://user-images.githubusercontent.com/20221896/105803339-36173f80-5fd8-11eb-9482-1f8fab289531.png)

however, python3-dev depends on that python3 package, so if you `apt install python3-dev`, you are basically installing a second copy of the python3 binary.

I would suggest building from the debian:buster-slim image, but maybe there are better ways.

#### 3. dev headers

furthermore, the dev headers are somewhat not needed in the running container. so maybe the `pip install` should go to a building stage, leaving the `site-packages` folder to the actual env

--------

I'll try to implement this, if anyone have better ideas, please let me know

**Comment by frankli0324** [comment]
update, since we are using buster base images now, `gevent` comes with a python wheel, that means that we needn't install build dependencies at all, it's prebuilt already. that makes things way easier.

**Comment by frankli0324** [comment]
seems effective, and runnable

```logs
➜  CTFd git:(master) ✗ docker images test
zsh: correct 'test' to 'tests' [nyae]? n
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
test         latest    5c1b22060779   3 seconds ago   485MB
➜  CTFd git:(master) ✗ docker images test2
REPOSITORY   TAG       IMAGE ID       CREATED              SIZE
test2        latest    fa99147e5524   About a minute ago   773MB
➜  CTFd git:(master) ✗ docker run -it --rm -p 7000:8000 test
zsh: correct 'test' to 'tests' [nyae]? n
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 75e8ab9a0014
 * Loaded module, <module 'CTFd.plugins.challenges' from '/opt/CTFd/CTFd/plugins/challenges/__init__.py'>
 * Loaded module, <module 'CTFd.plugins.dynamic_challenges' from '/opt/CTFd/CTFd/plugins/dynamic_challenges/__init__.py'>
 * Loaded module, <module 'CTFd.plugins.flags' from '/opt/CTFd/CTFd/plugins/flags/__init__.py'>
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
Starting CTFd
[2021-01-26 08:31:36 +0000] [1] [INFO] Starting gunicorn 20.0.4
[2021-01-26 08:31:36 +0000] [1] [INFO] Listening at: http://0.0.0.0:8000 (1)
[2021-01-26 08:31:36 +0000] [1] [INFO] Using worker: gevent
[2021-01-26 08:31:36 +0000] [11] [INFO] Booting worker with pid: 11
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
 * Loaded module, <module 'CTFd.plugins.challenges' from '/opt/CTFd/CTFd/plugins/challenges/__init__.py'>
 * Loaded module, <module 'CTFd.plugins.dynamic_challenges' from '/opt/CTFd/CTFd/plugins/dynamic_challenges/__init__.py'>
 * Loaded module, <module 'CTFd.plugins.flags' from '/opt/CTFd/CTFd/plugins/flags/__init__.py'>
[2021-01-26 08:31:39 +0000] [1] [INFO] Handling signal: winch
[2021-01-26 08:31:39 +0000] [1] [INFO] Handling signal: winch
[2021-01-26 08:31:39 +0000] [1] [INFO] Handling signal: winch
[2021-01-26 08:31:40 +0000] [1] [INFO] Handling signal: winch
[2021-01-26 08:31:40 +0000] [1] [INFO] Handling signal: winch
[2021-01-26 08:31:40 +0000] [1] [INFO] Handling signal: winch
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET / HTTP/1.1" 302 219 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /setup HTTP/1.1" 200 12716 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/css/fonts.min.css?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/css/main.min.css?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/css/core.min.css?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/js/core.min.js?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/js/helpers.min.js?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/js/pages/setup.min.js?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/js/vendor.bundle.min.js?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /themes/core/static/img/favicon.ico?d=c212d639 HTTP/1.1" 200 0 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /events HTTP/1.1" 302 219 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
172.17.0.1 - - [26/Jan/2021:08:31:56 +0000] "GET /setup HTTP/1.1" 200 12713 "http://localhost:7000/setup" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_1_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36"
```

**Comment by frankli0324** [comment]
I think this is resolved by the latest release 3.5.2
